### PR TITLE
PEFramework for loading GTA SA executable

### PR DESCRIPTION
**Pull request by quiret**
I wrote the PEFramework library to simplify using the PE file format. The library exposes all of the variables inside PE files. Please consider using it and providing feedback to me about things you would like to see improved.

**Steps which explain the enhancement**

1. Adds two dependencies (PEFramwork itself and some utilities in vendor/eirrepo)
2. Improves the executable loading process + better error handling
3. Modifies the current launcher code

**Current and suggested behavior**

It used ExecutableLoader before, It uses PEFramework now.

**Why would the enhancement be useful to most users**

Better error handling and supporting more PE formats.

**Screenshots, YouTube videos and GIFs**

--

**San Andreas Online Version:** v0.1.0-alpha.0
**OS and Version:** Windows 10

### CODE REVIEW

File: Client/Launcher/ExecutableLoader.cpp [hunk 1]
```diff
@@ -81,72 +140,62 @@ void ExecutableLoader::LoadImports(IMAGE_NT_HEADERS* ntHeader)
...
 
-        VirtualProtect(targetAddress, sizeOfData, PAGE_EXECUTE_READWRITE, &oldProtect);
-        memcpy(targetAddress, sourceAddress, sizeOfData);
+        sect.stream.Seek( 0 );
+        sect.stream.Read( targetAddress, rawSize );
     }
 }
 
-void ExecutableLoader::LoadSections(IMAGE_NT_HEADERS* ntHeader)
+void ExecutableLoader::LoadSections( void )
```
Comment (ifarbod) [inline Client/Launcher/ExecutableLoader.cpp]: Unnecessary 'void' keyword for functions which don't take a parameter.

File: Client/Launcher/ExecutableLoader.cpp [hunk 2]
```diff
@@ -81,72 +140,62 @@ void ExecutableLoader::LoadImports(IMAGE_NT_HEADERS* ntHeader)
...
-            nameTableEntry++;
             addressTableEntry++;
         }
-
-        descriptor++;
     }
 }
 
-void ExecutableLoader::LoadSection(IMAGE_SECTION_HEADER* section)
+void ExecutableLoader::LoadSection( PEFile::PESection& sect )
```
Comment (ifarbod) [inline Client/Launcher/ExecutableLoader.cpp]: Spaces between parentheses.

File: Client/Launcher/premake5.lua [hunk 3]
```diff
@@ -13,9 +13,14 @@ project "Launcher"
...
     links {
         "Utility",
-        "dbghelp", "psapi", "comctl32", "wininet", "winhttp", "Shlwapi"
+        "dbghelp", "psapi", "comctl32", "wininet", "winhttp", "Shlwapi",
+        "PEFramework"
     }
-
+    
+    includedirs {
+        "../../Vendor/eirrepo/"
```
Comment (ifarbod) [inline Client/Launcher/premake5.lua:21]: What does it requires from eirrepo?

**Review by ifarbod — CHANGES_REQUESTED** [review]
Fix these and I will merge it. Although it can be clang-format'd. 👍

File: Client/Launcher/premake5.lua (hunk 3 above)
Comment (quiret) [reply Client/Launcher/premake5.lua:21]: It requires rwlist.hpp, MemoryRaw.h, MemoryUtils.h and MemoryUtils.stream.h . Those are being created by PEFramework.

**Review by ifarbod — APPROVED** [review]
Thanks for adhering to our coding conventions.

**Comment by quiret** [comment]
I will adhere in the future aswell :) If this gets merged I will contribute great tech(tm) from time to time.

**Comment by ifarbod** [comment]
@AppVeyorBot Nice.

**Pull request closed without merging**